### ISSUE TITLE
Authentication failed Error on large files and no resume

### ISSUE THREAD

**Issue by yonjah** [ncw/rclone#820]
rclone v1.33
Linux Fedora 23 64 bit
OneDrive
When filing an issue, please include the following information if
possible as well as a description of the problem.  Make sure you test
with the [latest beta of rclone](http://rclone.org/downloads/).
rclone copy local to remote

Log >>>

```
2016/10/23 22:06:24 
Transferred:   210 MBytes (59.686 kBytes/s)
Errors:                 0
Checks:              1029
Transferred:            0
Elapsed time:    1h0m2.8s
Transferring:
 * ...dir/large_file.gz: 36% done. avg:   69.9, cur:    2.3 kByte/s. ETA: 45h28m38s

2016/10/23 22:07:24 
Transferred:   210 MBytes (58.708 kBytes/s)
Errors:                 0
Checks:              1029
Transferred:            0
Elapsed time:    1h1m2.8s
Transferring:
 * ...dir/large_file.gz: 36% done. avg:   68.5, cur:    0.0 kByte/s. ETA: 2486h39m29s

2016/10/23 22:07:35 folder/dir/large_file.gz: Cancelling multipart upload
2016/10/23 22:07:37 remote: Saving new token in config file
2016/10/23 22:07:37 folder/dir/large_file.gz: Failed to copy: unauthenticated: Authentication failed
2016/10/23 22:07:37 folder/dir/large_file.gz: Removing failed copy
2016/10/23 22:07:37 folder/dir/large_file.gz: Failed to remove failed copy: invalidRequest: API not found
2016/10/23 22:07:37 folder/dir/next_file.gz: Size and modification time the same (differ by 0s, within tolerance 1s)
2016/10/23 22:07:37 folder/dir/next_file.gz: Unchanged skipping
```

I have a very slow and unstable internet connection. when I try to upload large files (>500MB ) I usually get  an Authentication Error after the file is being upload for a large period of time.
I guess this is because it takes so long the authentication tokens are invalid.

When this error occurs rclone reauthenticates but skip the file and does not resume the upload.
This is not only very frustrating (when it happens at 97%) but make it impossible to upload very large files (> 1,5G) since with my connection speed they will always eventually fail.

I'm not sure if onedrive (and other services) allows to re authenticate and continue the upload (or if it possible to somehow refresh the token before it happens) but it would be a great improvement to have.

**Comment by ncw** [comment]
Amazon Drive had the same problem which was fixed in 2ebeed67531263294542834ea8edc7bb14334b09

I can add the same fix to onedrive.

I should probably factor the logic out and stick it in oauthutil then use it in all the oauth based remotes...

**Comment by mennojansz** [comment]
I'm having the same issue with rclone 1.35 on OneDrive and large file uploads.

**Comment by ncw** [comment]
I've attempted to fix this here.  Please have a test and re-open the ticket if it doesn't work!

Thanks Nick

http://beta.rclone.org/v1.35-52-g7470255/ (uploaded in 15-30 mins)

**Comment by yonjah** [comment]
I'm still seeing failures.
It seems like the token is being refreshed but the upload gets canceled -
```
ion time the same (differ by -759.707ms, within tolerance 1s)
2017/02/01 19:09:29 Pictures/872187382178.jpg: Unchanged skipping
2017/02/01 19:09:30 Documents/large_file.zip: Starting multipart upload
2017/02/01 19:09:31 Documents/large_file.zip: Uploading segment 0/1835489467 size 10485760
2017/02/01 19:11:39 Documents/large_file.zip: Uploading segment 10485760/1835489467 size 10485760
2017/02/01 19:13:49 Documents/large_file.zip: Uploading segment 20971520/1835489467 size 10485760
2017/02/01 19:16:04 Documents/large_file.zip: Uploading segment 31457280/1835489467 size 10485760
2017/02/01 19:18:15 Documents/large_file.zip: Uploading segment 41943040/1835489467 size 10485760
2017/02/01 19:20:26 Documents/large_file.zip: Uploading segment 52428800/1835489467 size 10485760
2017/02/01 19:22:34 Documents/large_file.zip: Uploading segment 62914560/1835489467 size 10485760
2017/02/01 19:24:42 Documents/large_file.zip: Uploading segment 73400320/1835489467 size 10485760
2017/02/01 19:26:50 Documents/large_file.zip: Uploading segment 83886080/1835489467 size 10485760
2017/02/01 19:28:58 Documents/large_file.zip: Uploading segment 94371840/1835489467 size 10485760
2017/02/01 19:31:06 Documents/large_file.zip: Uploading segment 104857600/1835489467 size 10485760
2017/02/01 19:33:15 Documents/large_file.zip: Uploading segment 115343360/1835489467 size 10485760
2017/02/01 19:35:23 Documents/large_file.zip: Uploading segment 125829120/1835489467 size 10485760
2017/02/01 19:37:31 Documents/large_file.zip: Uploading segment 136314880/1835489467 size 10485760
2017/02/01 19:39:43 Documents/large_file.zip: Uploading segment 146800640/1835489467 size 10485760
2017/02/01 19:41:52 Documents/large_file.zip: Uploading segment 157286400/1835489467 size 10485760
2017/02/01 19:44:01 Documents/large_file.zip: Uploading segment 167772160/1835489467 size 10485760
2017/02/01 19:46:12 Documents/large_file.zip: Uploading segment 178257920/1835489467 size 10485760
2017/02/01 19:48:25 Documents/large_file.zip: Uploading segment 188743680/1835489467 size 10485760
2017/02/01 19:50:38 Documents/large_file.zip: Uploading segment 199229440/1835489467 size 10485760
2017/02/01 19:52:48 Documents/large_file.zip: Uploading segment 209715200/1835489467 size 10485760
2017/02/01 19:54:59 Documents/large_file.zip: Uploading segment 220200960/1835489467 size 10485760
2017/02/01 19:57:14 Documents/large_file.zip: Uploading segment 230686720/1835489467 size 10485760
2017/02/01 19:59:36 Documents/large_file.zip: Uploading segment 241172480/1835489467 size 10485760
2017/02/01 20:01:42 One drive root 'backup/Private': Token expired - 1 uploads in progress - refreshing
2017/02/01 20:01:45 remote: Saved new token in config file
2017/02/01 20:01:57 One drive root 'backup/Private': Token refresh successful
2017/02/01 20:02:30 Documents/large_file.zip: Cancelling multipart upload
2017/02/01 20:02:31 Documents/folder/combo(27): Size and modification time the same (differ by 0s, within tolerance 1s)
2017/02/01 20:02:31 Documents/folder/combo(27): Unchanged skipping
```

I'm not sure what changed but I had at least one time where upload was successful after a token update -
```
2017/02/01 11:57:41 Pictures/IMG_20031012_112515.jpg: Unchanged skipping
2017/02/01 11:57:42 Pictures/large_video.mp4: Uploading segment 0/314967484 size 10485760
2017/02/01 11:59:50 Pictures/large_video.mp4: Uploading segment 10485760/314967484 size 10485760
2017/02/01 12:01:58 Pictures/large_video.mp4: Uploading segment 20971520/314967484 size 10485760
2017/02/01 12:04:05 Pictures/large_video.mp4: Uploading segment 31457280/314967484 size 10485760
2017/02/01 12:06:13 Pictures/large_video.mp4: Uploading segment 41943040/314967484 size 10485760
2017/02/01 12:08:22 Pictures/large_video.mp4: Uploading segment 52428800/314967484 size 10485760
2017/02/01 12:10:30 Pictures/large_video.mp4: Uploading segment 62914560/314967484 size 10485760
2017/02/01 12:12:38 Pictures/large_video.mp4: Uploading segment 73400320/314967484 size 10485760
2017/02/01 12:14:46 Pictures/large_video.mp4: Uploading segment 83886080/314967484 size 10485760
2017/02/01 12:16:54 Pictures/large_video.mp4: Uploading segment 94371840/314967484 size 10485760
2017/02/01 12:19:03 Pictures/large_video.mp4: Uploading segment 104857600/314967484 size 10485760
2017/02/01 12:21:11 Pictures/large_video.mp4: Uploading segment 115343360/314967484 size 10485760
2017/02/01 12:23:19 Pictures/large_video.mp4: Uploading segment 125829120/314967484 size 10485760
2017/02/01 12:25:27 Pictures/large_video.mp4: Uploading segment 136314880/314967484 size 10485760
2017/02/01 12:27:35 Pictures/large_video.mp4: Uploading segment 146800640/314967484 size 10485760
2017/02/01 12:29:44 Pictures/large_video.mp4: Uploading segment 157286400/314967484 size 10485760
2017/02/01 12:31:52 Pictures/large_video.mp4: Uploading segment 167772160/314967484 size 10485760
2017/02/01 12:34:00 Pictures/large_video.mp4: Uploading segment 178257920/314967484 size 10485760
2017/02/01 12:36:08 Pictures/large_video.mp4: Uploading segment 188743680/314967484 size 10485760
2017/02/01 12:38:16 Pictures/large_video.mp4: Uploading segment 199229440/314967484 size 10485760
2017/02/01 12:40:24 Pictures/large_video.mp4: Uploading segment 209715200/314967484 size 10485760
2017/02/01 12:42:32 Pictures/large_video.mp4: Uploading segment 220200960/314967484 size 10485760
2017/02/01 12:44:40 Pictures/large_video.mp4: Uploading segment 230686720/314967484 size 10485760
2017/02/01 12:46:48 Pictures/large_video.mp4: Uploading segment 241172480/314967484 size 10485760
2017/02/01 12:48:56 Pictures/large_video.mp4: Uploading segment 251658240/314967484 size 10485760
2017/02/01 12:48:57 remote: Saved new token in config file
2017/02/01 12:51:05 Pictures/large_video.mp4: Uploading segment 262144000/314967484 size 10485760
2017/02/01 12:53:13 Pictures/large_video.mp4: Uploading segment 272629760/314967484 size 10485760
2017/02/01 12:55:21 Pictures/large_video.mp4: Uploading segment 283115520/314967484 size 10485760
2017/02/01 12:57:29 Pictures/large_video.mp4: Uploading segment 293601280/314967484 size 10485760
2017/02/01 12:59:37 Pictures/large_video.mp4: Uploading segment 304087040/314967484 size 10485760
2017/02/01 13:01:45 Pictures/large_video.mp4: Uploading segment 314572800/314967484 size 394684
2017/02/01 13:01:52 Pictures/large_video.mp4: Copied (new)
2017/02/01 13:01:52 Pictures/IMG_20031109_032756.jpg: Size and modification time the same (differ by -135.876ms, within tolerance 1s)
2017/02/01 13:01:52 Pictures/IMG_20031109_032756.jpg: Unchanged skipping
```

**Comment by ncw** [comment]
OK I'll look at this some more...

**Comment by yonjah** [comment]
Ok I think I figured it out.
It seems like if uploading a part started but token has expired during the upload one drive will reject the part since the token is expired this error will stop the upload from being retried and the multipart session will be destroyed.

I submited a pull request but its my first time writing go code so there might be some improvements.
Instead of only checking the error code I'm checking that the error explicitly mention the token is expired I guess this can be simplified by retrying on any 401 error but then if the auth has a more serious error it will take it longer to fail.

I also found a bug that upload will not be able to actually retried since after the first attempt the Reader index will be at the end of the buffer and it need to be reseted
